rmw_fastrtps_shared_cpp/src/TypeSupport_impl.cpp

```diff
@@ -123,4 +141,450 @@ std::function<uint32_t()> TypeSupport::getSerializedSizeProvider(void * data)
   return ser_size;
 }
 
+// TODO(iuhilnehc-ynos): add the following content into new files named TypeObject?
+using CompleteStructType = eprosima::fastrtps::types::CompleteStructType;
+using CompleteStructMember = eprosima::fastrtps::types::CompleteStructMember;
+using MinimalStructType = eprosima::fastrtps::types::MinimalStructType;
+using MinimalStructMember = eprosima::fastrtps::types::MinimalStructMember;
+using SerializedPayload_t = eprosima::fastrtps::rtps::SerializedPayload_t;
+using TypeNamesGenerator = eprosima::fastrtps::types::TypeNamesGenerator;
+using TypeIdentifier = eprosima::fastrtps::types::TypeIdentifier;
+using TypeObject = eprosima::fastrtps::types::TypeObject;
+using TypeObjectFactory = eprosima::fastrtps::types::TypeObjectFactory;
+
+const rosidl_message_type_support_t *
+get_type_support_introspection(
+  const rosidl_message_type_support_t * type_supports)
+{
+  const rosidl_message_type_support_t * type_support =
+    get_message_typesupport_handle(
+    type_supports, rosidl_typesupport_introspection_c__identifier);
+  if (nullptr == type_support) {
+    rcutils_error_string_t prev_error_string = rcutils_get_error_string();
+    rcutils_reset_error();
+
+    type_support =
+      get_message_typesupport_handle(
+      type_supports,
+      rosidl_typesupport_introspection_cpp::typesupport_identifier);
+    if (nullptr == type_support) {
+      rcutils_error_string_t error_string = rcutils_get_error_string();
+      rcutils_reset_error();
+      RMW_SET_ERROR_MSG_WITH_FORMAT_STRING(
+        "Type support not from this implementation. Got:\n"
+        "    %s\n"
+        "    %s\n"
+        "while fetching it",
+        prev_error_string.str, error_string.str);
+      return nullptr;
+    }
+  }
+
+  return type_support;
+}
+
+template<typename MembersType>
+inline std::string
+_create_type_name(
+  const MembersType * members)
+{
+  if (!members) {
+    return std::string();
+  }
+
+  std::ostringstream ss;
+  std::string message_namespace(members->message_namespace_);
+  std::string message_name(members->message_name_);
+  if (!message_namespace.empty()) {
+    ss << message_namespace << "::";
+  }
+  ss << "dds_::" << message_name << "_";
+  return ss.str();
+}
+
+typedef std::pair<const TypeIdentifier *, std::string> MemberIdentifierName;
+
+template<typename MembersType>
+MemberIdentifierName GetTypeIdentifier(const MembersType * member, uint32_t index, bool complete);
+
+template<typename MembersType>
+const TypeObject * GetCompleteObject(
```

Comment:
Why is all this magic needed?
If xtypes would be enabled, would be need this?

Reply:
We need `rosidl_typesupport_introspection` to generate the xtypes information (a.k.a TypeObject/TypeIdentifier) so the content filter logic inside Fast DDS can calculate the offsets in the serialized payload to the fields of the type used in the filter expression.

So, in a way, what we are doing is *enabling* xtypes.

Reply:
Side note: we didn't do this on `rosidl_typesupport_fastrtps` because it is also used on `rmw_connextdds`, and it only depends on Fast CDR.

Also because supporting content filter on `rmw_fastrtps_dynamic_cpp` would also require the TypeObject generation.

That's why we put the code on the common package for both (i.e. `rmw_fastrtps_shared_cpp`)

Reply:
To see if I understand correctly, the generated code of `rosidl_typesupport_fastrtps` doesn't provide a typeobject, and we're dynamically creating one from the instronspection typesupport.

That seems okay, though it sounds like something that would better be done in the generated code.
If there would be a common API for generating type objects for both fastrtps and connext maybe the typeobject could also be provided by the generated code.

Reply:
Correct. But there's no common API, and `rmw_fastrtps_dynamic_cpp` depends only on `rosidl_typesupport_introspection`.